testing/test_hellsemble/hellsemble.py

```diff
@@ -351,12 +364,89 @@ def evaluate_hellsemble(
         }
 
         if callable(self.metric):
-            y_pred_proba = self.predict_proba(X)[:, 1]
-            return self.metric(y, y_pred_proba)
+            y_pred = self.predict(X)
+            return self.metric(y, y_pred)
 
         if self.metric == "roc_auc":
             y_pred_proba = self.predict_proba(X)[:, 1]
             return metrics_map["roc_auc"](y, y_pred_proba)
 
         y_pred = self.predict(X)
         return metrics_map[self.metric](y, y_pred)
+
+    def get_progressive_scores(
+        self, X: Union[np.ndarray, pd.DataFrame], y: Union[np.ndarray, pd.Series]
+    ):
+        """
+        Evaluates hellsemble based on primary metric used
+        for training in a greedy mode. Calculates the progressive scores of Hellsemble with each added model.
+
+        Args:
+            X (pd.DataFrame | np.ndarray): Feature matrix.
+            y (np.ndarray | pd.Series): Target vector
+
+        Returns:
+            list[float]: List of metric scores for each step in the ensemble.
+        """
+        metrics_map = {
+            "roc_auc": roc_auc_score,
+            "accuracy": accuracy_score,
+            "balanced_accuracy": balanced_accuracy_score,
+            "f1": f1_score,
+        }
+        estimators_copy = self.estimators.copy()
+        scores = []
+        for i in range(1, len(estimators_copy) + 1):
+            self.estimators = estimators_copy[:i]
+            if callable(self.metric):
+                y_pred = self.predict(X)
+                scores.append(self.metric(y, y_pred))
+            elif self.metric == "roc_auc":
+                y_pred_proba = self.predict_proba(X)[:, 1]
+                scores.append(metrics_map["roc_auc"](y, y_pred_proba))
+            else:
+                y_pred = self.predict(X)
+                scores.append(metrics_map[self.metric](y, y_pred))
+        return scores
+
+    def evaluate_routing_model(
```

Comment:
can be done with vectors in sth like :
```
preds = np.vstack([f(x) for f in fs])
np.abs(preds - y_true)
...
```

Two things to consider:
* afaik `score` is in [0, 1] so there may be ties but are not handled
* I am not sure if this evaluation is correct (or consistent with our idea); when model `i` handles observation `x_j`, it's still okay that by acciedent model `>i` is better to `x_j` so I have no idea if we should count it as a lack of the bad accuracy; I wish @azoz01 would have some insight here

Reply:
Both approaches to the evaluation are ok. The current one, in which we look for precisely one model (i-th), is beneficial since it shows how well the model learned the task. The second shows the performance of the model in an actual application. We should consider both metrics: first, for theoretical considerations and debugging, and second, for objective quality assessment.

Reply:
not sure exactly what the second method is, what it should do

Reply:
basically, when we train the `i`-th model, we expect that it should be better than all `<i` models in its area of application (because previous ones failed in this area) but we do not really know if sth similar holds for models `>i` (should be but our approach do not assume it). So if the router for observation `x` chooses model `i`, we would like to check if `i` is better than all `<i` **but not `>i`**. I now it's kinda tricky, also it's not really important so we may ignore it for now and discuss it during the meeting

Reply:
doesnt the greedy approach check against all i?

Reply:
I'd like to recall the `score > best_score` here - ties are totally okay IMO and now we get correct routing only if the model is strictly the best. Maybe calculate sth like `np.abs(y[idx] - y_hatpidx])` and check if for `routing_predictions[idx]` it's equal to `min`

Reply:
added below